pkg/storage/cacher.go

```diff
@@ -307,21 +381,53 @@ func (c *Cacher) Codec() runtime.Codec {
 	return c.storage.Codec()
 }
 
+func (c *Cacher) indexValues(event *watchCacheEvent) ([]string, bool) {
+	if c.indexFunc == nil {
+		return nil, false
+	}
+	// TODO: Currently we assume that only one index exists, thus
+	// exactly 1 value will be returned.
+	matchValues := c.indexFunc(event.Object)
+	if event.PrevObject == nil {
+		return []string{matchValues[0].Value}, true
+	}
+	prevMatchValues := c.indexFunc(event.Object)
+	if matchValues[0].Value == prevMatchValues[0].Value {
+		return []string{matchValues[0].Value}, true
+	}
+	return []string{matchValues[0].Value, prevMatchValues[0].Value}, true
+}
+
 func (c *Cacher) processEvent(event watchCacheEvent) {
+	indexValues, supported := c.indexValues(&event)
+
 	c.Lock()
 	defer c.Unlock()
-	for _, watcher := range c.watchers {
+	// Iterate over "allWatchers" no matter what the indexing function is.
+	for _, watcher := range c.watchers.allWatchers {
 		watcher.add(event)
 	}
+	if supported {
+		// Iterate over watchers interested in the given values of the index.
+		for _, indexValue := range indexValues {
+			for _, watcher := range c.watchers.valueWatchers[indexValue] {
+				watcher.add(event)
+			}
+		}
+	} else {
+		// Iterate over watchers interested in exact values for all values.
```

Comment:
and by definition, if you failed to get a value for the index doesn't that mean that none of these watchers will want to see it?